### ISSUE TITLE
primary_demand_of_biodiesel should return non-zero result for biomass branch

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etmodel#2916]
`V(energy_export_biodiesel, primary_demand)` gives a result of 0.0 locally but returns 53 PJ on BETA.

@antw could you help figure out why this happens? The `energy_production_biodiesel` still is part of the `primary_energy_demand` groups so we are a bit lost as to why the `primary_demand` methods seems to have stopped working for this carrier.

**Comment by antw** [comment]
On beta, energy_production_biodiesel is a "right dead end": it has no input links. On the biomass branch, there are three input links to the production node. The primary demand calculation traverses to the far-right of the graph and stops there; it does *not* terminate early when reaching a node in the `primary_energy_demand` group.

I expect`primary_demand` is being confused with `final_demand`, which *does* terminate early when encountering a node in "final_demand_group".

Is it possible to add the three energy_production_biodiesel inputs to the primary demand group instead?

**Comment by ChaelKruip** [comment]
> it does not terminate early when reaching a node in the primary_energy_demand group.

Ahh. I then misinterpreted some parts of the recursive method in ETEngine where it seems (to me) that it either stops at a `domestic_dead_end` or a `primary_energy_demand` node. E.g.,

```
return nil if !domestic_dead_end? || !primary_energy_demand?
```

> I expect `primary_demand` is being confused with final_demand, which does terminate early when encountering a node in "final_demand_group".

Right! 

> Is it possible to add the three energy_production_biodiesel inputs to the primary demand group instead?

That is indeed what I am planning to do but I wanted to understand the current situation first 😄 

Thanks! Closing.

**Comment by antw** [comment]
> I then misinterpreted some parts of the recursive method in ETEngine where it seems that it either stops at a domestic_dead_end or a primary_energy_demand node.
>
> E.g., `return nil if !domestic_dead_end? || !primary_energy_demand?`

Indeed! The `return nil` means the opposite of what you expected: *"if this is not a domestic dead-end (border between domestic and abroad or a right dead-end) OR not a primary energy demand node, continue traversing right"*.

It could be rewritten as:

```ruby
if domestic_dead_end? && primary_energy_demand?
  # do calculation
else
  # continue traversing
end
```

**Comment by michieldenhaan** [comment]
> Is it possible to add the three energy_production_biodiesel inputs to the primary demand group instead?

Just to be sure, we aren't going to add biosources like corn and wastefats to the primary demand group right? That would mean that the ETM's definition of primary energy demand is no longer in line with IEA/CBS etc.

**Comment by michieldenhaan** [comment]
Just had a call about this with @ChaelKruip. He pointed out that corn and manure are currently already counted as primary energy demand in the ETM. In that sense, adding waste fats, rapeseed, palm oil etc. to the primary demand group would be in line with how the ETM currently works.

I think we should try to avoid this. This would mean that the ETM's definition of primary energy demand is different from international standards, making it hard to compare ETM primary demand with other models/reports. In addition, it puts a large research burden on ensuring that the production shares (how much biodiesel comes from rapeseed, how much from waste fats etc.) are correct for each country/year.

I think we should split our energy calculation from our biomass calculation. I.e.:
- The primary energy demand calculation should be performed on all nodes in the primary energy demand group, regardless of whether they are a ``domestic_dead_end`` or not. In that way we can ensure that the ETM's primary carriers conform to international standards. I think we already do this in some sense for electricity from renewable sources.
- The calculation of biosources, land use from biomass, CO2 from biomass, FCE etc. should be able to traverse beyond that and e.g. stop when a node is in the biosources group (or a right dead end)

Curious to hear your views on this @AlexanderWirtz, @antw!

**Comment by antw** [comment]
@michieldenhaan If I understand correctly, what you're proposing is:

1. `demand_of_bio_resources_including_abroad` should stay as it is today.
2. `primary_demand` should stop recursing when it reaches a node in the "primary_energy_demand" group.

I have no strong feelings on this, so I will defer to everyone else's expertise on the "correctness" of this modelling.

On the tech side it's relatively easy to change, but there will be some fallout:

- I'll need to check what nodes may be affected by this, and get some idea of whether we'll have to adjust gqueries.
- `primary_co2_emission` will also likely need to be changed.
- We would need to figure out whether this impacts `primary_demand_of_carrier`.

I think this can be done and I expect the implementation to be fairly straightforward, but it will need plenty of testing.

**Comment by ChaelKruip** [comment]
> primary_co2_emission will also likely need to be changed.

I expect so yes. But how depends on how we define `primary_co2_emission`:
- The emission of the carrier at the end of the graph (right-most nodes). In this case, we would have to change the co2 calculation to keep recursing to the 'end'.
- The emission of the carrier in a specific group (which could be still the `primary_energy_demand` group or a new `co2_carriers` group). In this case, I imagine the recursive method to stop at nodes in this group rather than traversing to the 'end'.

> I think this can be done and I expect the implementation to be fairly straightforward, but it will need plenty of testing.

I agree!
